feat(rules): add IDV_CHECK operator and documentation

alonp99 opened the pull request:
- Implement IDV_CHECK operator for evaluating decision statuses in child workflows
- Enhance risk rules development guide with detailed operator creation steps
- Add unit tests for IDV_CHECK operator to validate expected behavior

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Expanded the rule engine with new risk screening operations that enable enhanced identity checks and company sanctions evaluations.
  - Introduced new constants and schemas for `IDV_CHECK` and `COMPANY_SANCTIONS_CATEGORIES` to enhance rule processing capabilities.
- **User Interface**
  - Removed the "Fill Placeholders" button for a more streamlined collection flow.
- **Documentation**
  - Introduced a comprehensive guide for developing and configuring risk rules.
- **Maintenance**
  - Updated dependencies and performed internal refinements to improve overall system stability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

alonp99 committed:
feat(rules): add IDV_CHECK operator and documentation

- Implement IDV_CHECK operator for evaluating decision statuses in child workflows
- Enhance risk rules development guide with detailed operator creation steps
- Add unit tests for IDV_CHECK operator to validate expected behavior

alonp99 committed:
test(rule-engine): add unit tests for company sanctions adverse media rule

- Introduce tests to validate behavior of company sanctions adverse media detection
- Implement handling for cases with and without adverse media data
- Update rule engine error message to include COMPANY_SANCTIONS_ADVERSE_MEDIA

alonp99 committed:
fix(workflows-plugins): individuals sanaction plugin fixes

alonp99 committed:
feat(rule-engine): add company sanctions adverse media operator

- Implement new operator for evaluating adverse media in company sanctions
- Create corresponding tests to validate the functionality and threshold logic
- Refactor existing rules to incorporate the new operator and ensure consistency

alonp99 committed:
refactor(tests): simplify schema definition in rule engine unit tests

- Remove unnecessary boolean check from schema
- Streamline schema structure for clarity

alonp99 committed:
refactor(company-sanctions): rename adverse media operator to categories

- Rename CompanySanctionsAdverseMedia to CompanySanctionsCategories
- Update all related imports, schemas, and tests accordingly
- Adjust evaluation logic to handle category-based thresholds

alonp99 commented on packages/common/src/rule-engine/operators/index.ts:
```diff
@@ -0,0 +1,46 @@
+import { OPERATION } from './enums';
```
Circular dependency fix when trying to create rules in a separate file - tried to avoid additional bloat to helpers.ts..

alonp99 committed:
fix(workflows): remove commented-out rule for company sanctions

- Eliminate outdated comment on company sanctions rule
- Ensure clarity and maintainability of the code

tomer-shvadron commented on packages/common/src/rule-engine/operators/company-sanctions-by-category.ts:
```diff
@@ -0,0 +1,93 @@
+                  .optional(),
+              }),
+            )
+            .min(1, 'Company sanctions data is required'),
```
I'm not sure about this line, the array is empty when there are no hits. WDYT?

jonathan-shemer replied:
nitpick 
```suggestion
            .notempty('Company sanctions data is required'),
```

tomer-shvadron commented on packages/common/src/rule-engine/operators/company-sanctions-by-category.ts:
```diff
@@ -0,0 +1,93 @@
+      throw new DataValueNotFoundError('pluginsOutput.companySanctions.data');
+    }
+
+    const sourceCategories: string[] = [];
+
+    companySanctions.forEach(sanction => {
+      if (sanction?.entity?.sources) {
+        sanction.entity.sources.forEach(source => {
+          if (source?.categories && Array.isArray(source.categories)) {
+            source.categories.forEach(category => {
+              sourceCategories.push(category);
+            });
+          }
+        });
+      }
+    });
+
+    return sourceCategories;
```
Can this section be replaced by `flatMap`?

tomer-shvadron commented on packages/common/src/rule-engine/operators/index.ts:
```diff
@@ -0,0 +1,46 @@
+import { OPERATION } from './enums';
+
+// Export base operators class
```
Please remove all of these unnecessary AI-generated comments :)

jonathan-shemer commented on packages/common/src/rule-engine/operators/company-sanctions-by-category.ts:
```diff
@@ -0,0 +1,93 @@
+      }),
+    });
+
+    const result = companySanctionsSchema.safeParse(data);
```
nitpick
why use `safeParse` if we alway throw an exception when `false`?

alonp99 replied:
To control the error message basically, but I can't check if it logs fine just with not safe parse

jonathan-shemer commented on packages/common/src/rule-engine/operators/company-sanctions-by-category.ts:
```diff
@@ -0,0 +1,93 @@
+
+    const companySanctions = get(objData, 'pluginsOutput.companySanctions.data');
+
+    if (!companySanctions || isEmpty(companySanctions)) {
```
is this going to happen when you've asserted this with `zod`?

jonathan-shemer commented on packages/common/src/rule-engine/operators/company-sanctions-by-category.ts:
```diff
@@ -0,0 +1,93 @@
+
+    const sourceCategories: string[] = [];
+
+    companySanctions.forEach(sanction => {
```
can this be included in the `zod` schema? also, asserting it's not empty

jonathan-shemer commented on packages/common/src/rule-engine/operators/company-sanctions-by-category.ts:
```diff
@@ -0,0 +1,93 @@
+        sanction.entity.sources.forEach(source => {
+          if (source?.categories && Array.isArray(source.categories)) {
+            source.categories.forEach(category => {
+              sourceCategories.push(category);
```
you may transform the category in the `zod` schema, like i've done in [https://github.com/ballerine-io/ballerine/commit/3e647efa51876bf1da8f5601d1b6c510f41ec780#diff-fa6303dd6f6948eec3bc426f3d9aaaf3eb8bde7e99431add2a317fb219eb3652R468](normalizedString), and even re-use the same transformation.

jonathan-shemer commented on packages/common/src/rule-engine/operators/company-sanctions-by-category.ts:
```diff
@@ -0,0 +1,93 @@
+    return sourceCategories;
+  }
+
+  evaluate(dataValue: string[], conditionValue: CompanySanctionsCategoriesParams): boolean {
+    if (!dataValue || dataValue.length === 0) {
+      return false;
+    }
+
+    const threshold = conditionValue?.threshold || 1;
+
+    const categoryCount = dataValue.filter(
+      category =>
+        typeof category === 'string' &&
+        category.toLowerCase() === conditionValue.category.toLowerCase(),
+    ).length;
+
+    return categoryCount >= threshold;
+  }
+}
```
applying more logic to the `zod` schema would allow evaluate to be the simplest
```typescript

  evaluate(dataValue: string[], conditionValue: CompanySanctionsCategoriesParams): boolean {
    const threshold = conditionValue?.threshold || 1;
    const categoryCount = dataValue.length;
    return categoryCount >= threshold;
  }

jonathan-shemer commented on services/workflows-service/src/rule-engine/core/test/idv-check.unit.test.ts:
```diff
@@ -0,0 +1,276 @@
+describe('IDV_CHECK operator', () => {
+  it('should pass when decision status is declined', async () => {
+    const mockData = {
+      childWorkflows: {
```
this is repeated a lot, i would have a helper method `createDecision(status: string, score: number)` and reuse it.
it's more readable that way

alonp99 replied:
Will handle in a new PR

alonp99 committed:
fix(workflows-plugins): individuals sanaction plugin fixes

alonp99 committed:
fix(slack): fix mm customer name

alonp99 committed:
refactor(CollectionFlowV2): remove unused placeholder filling button

- Remove the button responsible for filling placeholders from the UI
- Optimize the readability and maintainability of CollectionFlowV2 component

fix(rule-engine): change error message for missing decisions

- Modify error thrown in IdvCheck class for better clarity

fix(schemas): update key property in rule schema to optional

- Change key property in the rule schema to optional to enhance flexibility

alonp99 committed:
fix(slack): fix mm customer name

alonp99 committed:
fix(workflows-plugins): individuals sanaction plugin fixes

alonp99 committed:
refactor(types): update array type annotations for consistency

- Change type annotations from Array<T> to shorthand T[] where appropriate
- Enhance type consistency in function parameters and return types
- Improve code readability and maintainability by using preferred notations
